### ISSUE TITLE
Possible bug in df.update

### ISSUE THREAD

**Issue by isaacgg** [pandas-dev/pandas#29462]
#### Code Sample, a copy-pastable example if possible

```python
test_json = [{"_id": 'a', 'date': datetime.now()}, {"_id": 'b', 'date': datetime.now()}]
test_df = pd.DataFrame(test_json)

new_df = test_df.copy()
new_df["date"] = None
new_df.update(test_df)

print(test_df.head())
print(new_df.head())

```
#### Problem description

When using update function with datetime data, it is automatically converted to timestamp, which for me it seems like an abnormal behaviour. Code from above would output

  _id                       date
0   a 2019-11-07 15:50:06.072158
1   b 2019-11-07 15:50:06.072158
  _id                 date
0   a  1573141806072158000
1   b  1573141806072158000

#### Expected Output

  _id                       date
0   a 2019-11-07 15:50:06.072158
1   b 2019-11-07 15:50:06.072158
  _id                 date
0   a 2019-11-07 15:50:06.072158
1   b 2019-11-07 15:50:06.072158

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 61 Stepping 4, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 0.25.1
numpy            : 1.16.4
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.2.2
setuptools       : 41.0.1
Cython           : 0.29.13
pytest           : None
hypothesis       : None
sphinx           : 2.1.2
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.1
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.7.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
</details>

**Comment by susan-shu-c** [comment]
Hi, I was able to reproduce this result. This is due to `pandas.DataFrame.update` calling `expressions.where`. [source link](https://github.com/pandas-dev/pandas/blob/v0.25.3/pandas/core/frame.py#L5819).

From then it eventually calls `numpy.where` [documentation](https://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.where.html) which then eventually uses the Numpy `MaskedArray` type. [source link](https://github.com/numpy/numpy/blob/718c63f9d2956ac0a962f0795fce5d8e7a1a1a77/numpy/ma/core.py#L2706).

This seems to be a choice to use `numpy.where`, which causes the `datetime` type to be converted to unix time, which speeds up the computation. However feel free to correct me on that.

I'd suggest trying `pandas.to_datetime` [linked here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html) to convert them afterward (sometimes you have to divide the unix time by 1000 to get it to work), but I haven't tested on your example data yet, so feel free to try it.

**Comment by rhshadrach** [comment]
Indeed, this appears to be an odd interaction with DatetimeArray and `np.where`.

````
a = np.asarray([None], dtype=np.object)
b = np.asarray(pd.arrays.DatetimeArray(pd.Series([datetime.now()])))
cond = [False]
print(np.where(cond, a, b))
````

gives `[1595782471507905000]`; whereas

````
a = np.asarray([None], dtype=np.object)
b = np.asarray([datetime.now()], dtype=np.object)
cond = [False]
print(np.where(cond, a, b))
````

gives `[datetime.datetime(2020, 7, 26, 16, 53, 4, 806281)]`

**Comment by jbrockmendel** [comment]
i see the expected behavior on main (looks like a np.where call got changed to use Series.where somewhere along the line).  Could use a test (first check to see if one already exists)

**Comment by takesanocean** [comment]
Not reproducible for me either. Will check if there is a test already for this and will create one if none 

`take`

**Comment by takesanocean** [comment]
@jbrockmendel 
Hi! I wrote a test and also checked `dtypes` between these two dataframes. I see that the “date” column has `dtype=object` in the `new_df` data frame instead of `dtype=np.datetime[64]` (after updating DataFrame in this scenario). 
So it's not converting to timestamp anymore yeah, but looks like it's not ideal still.

```python
other = df.copy()
other["date"] = None # setting column to None is essential here to see an abnormal behaviour
other.update(df)
```
```output
DataFrames
df:
  _id                       date
0   a 2025-11-10 11:01:28.938296
1   b 2025-11-10 11:01:28.938299

other:
  _id                        date
0   a  2025-11-10 11:01:28.938296
1   b  2025-11-10 11:01:28.938299

DataFrames' dtypes
df:
_id                str
date    datetime64[us]
dtype: object

other:
_id        str
date    object
dtype: object
```

I will create now a small fix in `update()` function for this case to ensure `datetime64` is used in the copy if that makes sense to you too.

UPD
checked and it's more involved when I thought. Trying to figure out the best approach to this issue still

**Comment by roeimed0** [comment]
Hi, it looks like this hasn't had a PR in a few months. I'd like to pick this up if no one is actively working on it, it will be my first contribution

**Comment by roeimed0** [comment]
@jbrockmendel 
after reviewing update() function and this bug, I came to the following conclusions: 
The original issue of datetime values turning into integer timestamps is fixed.

dtype issue remains; when a column is set to None it is converted to object dtype, update() restores the correct values, but the column stays as object dtype instead of reverting to the original dtype. This affects other dtypes like int64 and str.

Looking at the docstring, there's are no design decisions about dtype preservation between self and other when comparing object column and a dtype column.

The current issue is a design question, not a bug: Should update() preserve self's dtype as object?

I will add a regression test to make sure the bug doesn't return, but the design question will affect the test- should I implement the test without checking dtypes or should I attempt to change update to take the precise dtype from the reference dataframe if self has a completely empty column?
